File: server/src/Models/DMToolkitModels/Entities/Effect.cs
using DMToolkit.API.Models.DMToolkitModels.Definitions;
using DMToolkit.API.Models.DMToolkitModels.Items.Definitions;
using DMToolkit.API.Models.DMToolkitModels.JoinTables;
using SharedModels.Enums;

namespace DMToolkit.API.Models.DMToolkitModels.Entities;

public class Effect
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public string? Title { get; set; } = null;
    public string? Description { get; set; } = null;
    public OriginType OriginType { get; set; } = OriginType.Other;
    public DataType DataType { get; set; } = DataType.Other;
    public Dictionary<string, object> Data { get; set; } = new();
    public ICollection<FeatDefinitionEffect> FeatDefinitionEffects { get; set; } = new List<FeatDefinitionEffect>();
    public ICollection<Spell> Spells { get; set; } = new List<Spell>();
    public ICollection<ConditionDefinition> ConditionDefinitions { get; set; } = new List<ConditionDefinition>();
    public ICollection<WeaponDefinition> WeaponDefinitionProperties { get; set; } = new List<WeaponDefinition>();
    public ICollection<WeaponDefinition> WeaponDefinitionMasteries { get; set; } = new List<WeaponDefinition>();
}

File: flutter-client/dm_toolkit/lib/models/dm_toolkit/entities/effect.dart
import 'dart:developer';

import 'package:dm_toolkit/enums/data_type.dart';
import 'package:dm_toolkit/enums/origin_type.dart';

class Effect {
  final String id;
  final String? title;
  final String? description;
  final OriginType originType;
  final DataType dataType;
  final Map<String, dynamic> data;

  Effect({
    required this.id,
    required this.title,
    required this.description,
    required this.originType,
    required this.dataType,
    required this.data
  });

  factory Effect.fromJson(Map<String, dynamic> json) {
    try {
      log('Data');
      return Effect(
        id: json['id'] as String, 
        title: json['title'] as String?,
        description: json['description'] as String?,
        originType: OriginType.values[json['originType'] as int], 
        dataType: DataType.values[json['dataType'] as int], 
        data: json['data'] as Map<String, dynamic>
      );
    } catch (e) {
      throw FormatException('Effect model is invalid.', e);
    }
  }
}